source-dropbox/oauth2.go

```diff
@@ -0,0 +1,72 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	pf "github.com/estuary/flow/go/protocols/flow"
+)
+
+func GetScopes() []string {
+	return []string{
+		"files.metadata.read",
+		"files.content.read",
+		// "account_info.read",  // Not sure if we need this
+	}
+}
+
+// Most of this code was converted from the Google Auth specification updating with
+// the info on https://developers.dropbox.com/oauth-guide
+func OAuth2Spec() *pf.OAuth2 {
+	// Docs: https://www.dropbox.com/developers/documentation/http/documentation#oauth2-authorize
+	AUTH_URL_TEMPLATE_FORMAT_STR := "https://www.dropbox.com/oauth2/authorize?token_access_type=offline&client_id={{#urlencode}}{{{ client_id }}}{{/urlencode}}&response_type=code&redirect_uri={{#urlencode}}{{{ redirect_uri }}}{{/urlencode}}&state={{{ state }}}&scope={{#urlencode}}%s{{/urlencode}}"
+
+	scopes := GetScopes()
+
+	spec := pf.OAuth2{
+		Provider:                  "dropbox",
+		AuthUrlTemplate:           fmt.Sprintf(AUTH_URL_TEMPLATE_FORMAT_STR, strings.Join(scopes, " ")),
+		AccessTokenUrlTemplate:    "https://api.dropboxapi.com/oauth2/token",
+		AccessTokenHeadersJsonMap: map[string]json.RawMessage{"Content-Type": json.RawMessage(`"application/x-www-form-urlencoded"`)},
+		AccessTokenBody:           "client_id={{{ client_id }}}&client_secret={{{ client_secret }}}&code={{{ code }}}&grant_type=authorization_code&redirect_uri={{{ redirect_uri }}}",
+		AccessTokenResponseJsonMap: map[string]json.RawMessage{
+			"access_token": json.RawMessage(`"/access_token"`),
+			"token_type":   json.RawMessage(`"/token_type"`),
+			"expires_in":   json.RawMessage(`"/expires_in"`),
+			"scope":        json.RawMessage(`"/scope"`),
+			"uid":          json.RawMessage(`"/uid"`),
+			"account_id":   json.RawMessage(`"/account_id"`),
+		},
+	}
+
+	return &spec
+}
+
+type Credentials struct {
+	AccessToken  string `json:"access_token"`
+	TokenType    string `json:"token_type"`
+	ExpireTime   int    `json:"expires_in"`
+	Scope        string `json:"scope"`
+	UID          string `json:"uid"`
+	AccountID    string `json:"account_id"`
+	ClientID     string `json:"client_id"`
+	ClientSecret string `json:"client_secret"`
+}
+
+func (c *Credentials) Validate() error {
+	requiredProperties := [][]string{
+		{"AccessToken", c.AccessToken},
+	}
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+	return nil
+}
+
+func (c *Credentials) GetAccessToken(ctx context.Context) (string, error) {
+	return c.AccessToken, nil
```

Comment:
I don't think this is going to work, since the access token will expire in a short time. It would mean that a user would need to come back to the system and re-authenticate when the access token expired. I believe the default is 4 hours so that wouldn't work well at all. We need to be getting a long-lived refresh token, and using that to get a new access token every time the connector starts up.

In the [rclone source](https://github.com/rclone/rclone/blob/5bf70c68f1ed0255be8592abe428b9d50ce37280/backend/dropbox/dropbox.go#L410) the `Client` property of the dropbox config is set to an oauth client. This seems a bit hacky since that field is documented as "for testing purposes only", but it seems like the best way to make this work and since it is implemented that way in rclone I feel decent about it.

Relatedly, I think the `AccessTokenResponseJsonMap` and fields of `Credentials` needs to be re-worked. We should only need `ClientID`, `ClientSecret`, and `RefreshToken`. It should look very similar to the google oauth in that regard. The fact that both systems are using Oauth2 and yet this implementation looks so different should be a strong indicator to you that this one is not right.

Since I was a little hazy on how the oauth stuff worked for connectors, I pushed a branch to prove out what I am saying. Check it out here: https://github.com/estuary/connectors/commit/3c421fe6ff8ac96892c291fc8d0b1482d544b57e

I think we should make updates along these lines to use the refresh token.